lighthouse-core/gather/computed/main-resource.js

```diff
@@ -0,0 +1,27 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const ComputedArtifact = require('./computed-artifact');
+const HTTP_REDIRECT_CODE_LOW = 300;
+const HTTP_REDIRECT_CODE_HIGH = 399;
+
+class MainResource extends ComputedArtifact {
+  get name() {
+    return 'MainResource';
+  }
+
+  /**
+   * @param {!Array<!WebInspector.NetworkRequest>} networkRecords
+   * @return {?WebInspector.NetworkRequest}
```

Comment:
when can this be null? What should a caller assume in those cases?

Reply:
I assumed it could happen in some edge cases (e.g. redirect loop? ssl error page?), but maybe I'm too paranoid? The audit that uses this artifact, in such case, will fail with message 'Invalid MainResouce or its status code.'.

Reply:
> I assumed it could happen in some edge cases (e.g. redirect loop? ssl error page?), but maybe I'm too paranoid?

Didn't mean to imply it was overkill; I think it's just the right amount of kill :) I was hoping we could enumerate possible failure cases so we could put more specific docs on the computed artifact (and in the audit `debugString`).

In the two cases I don't believe we should ever get as far as audits running, so I wonder if we should throw if nothing is found. That would allow outputting the exact error case as a string without any audit that consumes it having to detect those cases and outputting their own `debugString`

Reply:
@patrickhulce has strong feelings about Lighthouse errors vs site errors, but in this case I think we can say that if we don't have a main resource record it's either a pretty bad Lighthouse bug or a pretty bad Chrome bug, so I'm personally ok with throwing in this case (with specific info in error message so we can debug later)

Reply:
Makes sense! I've changed it to throw when main resource cannot be identified and adjusted tests accordingly.